### DISCUSSION TITLE
SPI transmit only

### DISCUSSION THREAD

**Discussion by robzed** [tabemann/zeptoforth#73] [Q&A] [answered]
I'm working with an LCD that you can only transmit to:

    1 lcd_clk spi-pin \ SPI1 SCK
    1 lcd_din spi-pin \ SPI1 TX
    1 master-spi
    10000000 1 spi-baud!
    8 1 spi-data-size!
   false false 1 motorola-spi      
   1 enable-spi

When transmitting a byte can I just do:
  ( n -- ) 1 >spi 

Or do I need to drop the incoming buyte?
  ( n -- ) 1 >spi  1 spi> drop

**Comment by robzed** [comment]
Also I write 25600 bytes to >SPI and it takes almost 0.5s ... 

lcd_buf_size . 25600  ok

\ Write a set of bytes to the SPI port
: spi_write ( address n -- )
  0 do 
    dup C@ 1 >spi 1+ 
  loop drop
;

It appears to take 458ms?
(mecrisp using bit bashing took 83ms - haven't timed zeptforth with bitbashing)

: lcd.display ( -- )
  0 lcd_height 1- 0 lcd_width 1- lcd.SetWindows       
  1 lcd_dc pin! 
  0 lcd_cs pin!
 us-counter-lsb
  lcd_buffer lcd_buf_size spi_write
 us-counter-lsb swap - u. cr

  1 lcd_cs pin!
;
: test_fill
 lcd.init
 BLUE lcd.fill lcd.display
 1000 ms
 RED lcd.fill lcd.display
 1000 ms
 GREEN lcd.fill lcd.display
 1000 ms
 BLACK lcd.fill lcd.display
;

test_fill 458064 
458147 
457919 
457802 
 ok

**Comment by robzed** [comment]
Software SPI (bitbash) version in Zeptoforth 

test_fill 411605 
411624 
411907 
409925

**Reply by robzed** [reply]
Maybe I should try buffer>spi

**Reply by tabemann** [reply]
I would definitely say try `buffer>spi` - just using `>spi` and `spi>` (they need to pair up) has a lot of extra overhead.

**Comment by tabemann** [comment]
BTW, have you used `spi-baud!` to set the baud for SPI?

**Reply by tabemann** [reply]
Forget that, I missed your earlier post.

**Comment by tabemann** [comment]
Just deleted my post, thinking I got something wrong in my math, but at second thought it looks right so I'm reposting this:

I wrote the following code:

```
spi import
timer import

\ 1 lcd_clk spi-pin \ SPI1 SCK
\ 1 lcd_din spi-pin \ SPI1 TX
1 master-spi
10000000 1 spi-baud!
8 1 spi-data-size!

false false 1 motorola-spi
1 enable-spi

25600 constant lcd_buf_size
lcd_buf_size buffer: lcd_buf

lcd_buf lcd_buf_size 0 fill

: spi_write ( address n -- )
0 do
dup C@ 1 >spi 1+ 1 spi> drop
loop drop
;

: time-it ( -- bitrate )
  us-counter-lsb { start }
  lcd_buf lcd_buf_size spi_write
  us-counter-lsb { end }
  end start - 1000000 swap / lcd_buf_size * 8 *
;

: time-it-again ( -- bitrate )
  us-counter-lsb { start }
  lcd_buf lcd_buf_size 1 buffer>spi
  us-counter-lsb { end }
  end start - 1000000 swap / lcd_buf_size * 8 *
;
```

and got the following result:

```
time-it-again . 1024000  ok
time-it . 614400  ok
```

i.e. when using `>spi` paired with `spi>` I got 614400 b/s and when I used `buffer>spi` I got 1024000 b/s.

**Reply by robzed** [reply]
Ok, so I must have a spi> and drop go every >spi

**Comment by tabemann** [comment]
I wonder if there's an issue with the clock scalers in use, i.e. why your bitbashed version wasn't much faster than the version using `>spi`.

**Comment by tabemann** [comment]
Okay, I did some calculations and got the following:

```
10000000 find-spi-prescale . 2  ok
10000000 2 find-spi-postdiv . 7  ok
125000000,0 2,0 7,0 1,0 d+ f* f/ f. 7812500,0  ok
```

i.e. it is really trying to do 7812500 baud, yet the software cannot keep up with the full speed and is instead limited to 1024000 baud. I could potentially rewrite this code to use DMA or PIO or like.

**Comment by tabemann** [comment, marked as answer]
I managed to optimize `buffer>spi` and `spi>buffer` quite a bit, speeding up `buffer>spi` by approximately a factor of 4.5:

```
time-it-again . 4505600  ok
```

Specifically, it takes 43.977 ms to now send the data.

**Reply by tabemann** [reply]
The code for this is in both the `build` and `devel` branches but is not currently in any builds.

**Reply by robzed** [reply]
Wow you have been busy!!! I will try this out later

**Reply by robzed** [reply]
Unmodified buffer>spi I get around 216915 for my text (216ms) - which is about twice as fast as my >spi one.

Haven't tried the updated one yet

**Reply by robzed** [reply]
Times in microseconds for buffer copy: 43877 43788 43788 43762
Faster than mecrisp bitbash!

Buffer fill and write to screen (with coords) 46813 46722 46690 46722. 
That's about 21 frames per seconds.

**Reply by tabemann** [reply]
And that's without using DMA too - I tried using DMA in hopes of pushing the bandwidth to the maximum of what the SPI peripheral theoretically supports but ran into issues with it, so I had to fall back on a highly optimized tight loop instead.
